pkg/reconciler/buildrun/resources/remote_artifacts.go

```diff
@@ -0,0 +1,70 @@
+package resources
+
+import (
+	"fmt"
+	"strings"
+
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/shipwright-io/build/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	v1 "k8s.io/api/core/v1"
+)
+
+// outputDocumentFlag used with wget to specify an alternative artifact location.
+const outputDocumentFlag = "--output-document"
+
+// outputPathForBuildSource return the final output path to a given artifact based on BuildSource
+// attributes. Can be empty when alternative location is not set.
+func outputPathForBuildSource(source buildv1alpha1.BuildSource) string {
+	if source.HTTP == nil {
+		return ""
+	}
+	return source.HTTP.Path
+}
+
+// renderRemoteArtifactsDownloadScript returns a slice of commands, a shell script, based on informed
+// BuildSources slice. Scripting lines are bind together with "&&".
+func renderRemoteArtifactsDownloadScript(sources []buildv1alpha1.BuildSource) []string {
+	script := []string{}
+	for _, source := range sources {
+		if len(script) > 0 {
+			script = append(script, "&&")
+		}
+		cmd := []string{"wget", source.URL}
+		if filePath := outputPathForBuildSource(source); filePath != "" {
+			cmd = append(cmd, fmt.Sprintf("%s=%s", outputDocumentFlag, filePath))
```

Comment:
If the output directory, or some parts of it do not exist, I think `wget` will not create those and fail with the download. Should we add `mkdir -p` or similar here?

Reply:
Creating the directory path before downloading the payload is another use-case that pushes for a dedicated application to handle this logic. The new actor should take care of creating the directory structure, respecting authentication, UID/GID and more, before fetching the payload. Furthermore, we can consider parallelism and other investments for performance related subjects.

Reply:
OK I'm a little confused / need some help on this thread?

So, we have 
```
func outputPathForBuildSource(source buildv1alpha1.BuildSource) string {
	if source.HTTP == nil {
		return ""
	}
	return source.HTTP.Path
}
```

And I see this in the new doc:

```
Under `.spec.sources` we have the following attributes:

- `.name`: represents the name of resource, required attribute.
- `.type`: data source type, currently only `http` type is supported, required attribute.
- `.url`: universal resource location (URL), required attribute.
- `.http.path`: alternative path to save URL contents, can be either a full or relative path, optional attribute.

When downloading artifacts the process is executed in the same directory where the application source-code is located, by default `/workspace/source`, therefore when using relative path on `.http.path` mind the starting location.
```

Are you saying @otaviof that another piece of code outside of this PR will be needed for creating `source.HTTP.Path` ?

If so, I'm not sure we should be adding fields we don't have a viable "default implementation" for all cases.  In which case, perhaps we have to be more precise / restrictive in what we support "out of the box" for the field.

One possible way to restrict:  perhaps we tweak the last sentence of the doc to say, at least for now, we only support relative paths off of the source code location, in which case a `mkdir -p` could be more viable 

Or perhaps the destination dir location always has to be `/workspace/source` and we allow the use of `wget -O` to change the name of the file that we store the contents in.  i.e. the remote artifact is `foo.tar` but we let them call it `bar.tar`.

Or I'm just confused / missing the point as I said up top and I need further explanation ;-)

Reply:
Forgot, not that this has to dictate which way we go, but for the analogous function in openshift builds, making relative subdirs is hard coded perms wise to 

`err := os.MkdirAll(destDir, 0755)`

and there is not chown's done.  And we have not had any complaints so far.  And it only supports relative paths off of a predefined / hardcoded mount in the pod.

Reply:
> Are you saying @otaviof that another piece of code outside of this PR will be needed for creating `source.HTTP.Path` ?

I'm afraid that's currently the case already, Gabe. Here in this PR we have `wget` being employed to download, and store, the remote artifacts, that's "another piece of code".

However, the current PR is evolving to become a "shell script" generator backed in the operator code. I would like to address this concern by having a clear separation of concerns where our "custom `wget`" would deal with such environment changes, like creating the underlying directory path, and many others to surface.

Reply:
I'm going to make changes in this PR to only accept relative locations (to `/workspace/source`), since this is the only mount-point where the user is allowed to. I hope that's a good starting point until we can properly address the separation of concerns.

Thanks for the input, the field `.http.path` is indeed a sharp corner.

Reply:
So @otaviof @coreydaley @adambkaplan @sbose78 and myself discussed this collectively in a team meeting this morning (US time) / afternoon (Europe time).

I'll try to summarize where we landed:
- we went from only accepting relative paths for `.http.path` to  removing the introduction of this field from this PR; instead, it will be introduced when some of the next steps @otaviof alluded to here show up in PRs. For the sake of brevity, I'll call the upcoming work "pluggable download mechanism".
- where needed (if anywhere) @otaviof will look at what happens when multiple sources download to the same file name under `/workspace/sources`, and making sure that it is either mitigated or surfaced in a suitable way to the user.

Of course @otaviof @coreydaley @adambkaplan @sbose78  chime in if I've missed something that should be noted here.

Reply:
Folks, please consider latest changes.